### ISSUE TITLE
Support LLM-based Merging

### ISSUE THREAD

**Issue by drewmccormack** [drewmccormack/Forked#5]
With a 3-way merge, there is presumably enough information for a LLM, even a small one, to come up with a good merge of divergent text. This seems like a natural evolution of data models: rather than using traditional algorithms, when a conflict arises on complex data like human-readable text, an LLM would presumably merge it in a more natural way.

CRDTs are partially fulfilling this role now in the framework, ie, making more "human" choices when merging text. But even those are just an algorithm attempting to do what a human might. An LLM is a much more advanced algorithm which is trained at exactly this type of problem, and an embedded LLM on a device should be able to do a great job of merging conflicts in text.

This would be a good issue for someone who has some experience with embedded LLMs. The idea would be to add a new type of merger called a `LLMMerger`. For inspiration, see `TextMerger`. The macros would then be updated to support a new merge algorithm, eg, `@Merged(using: .smartTextMerge)`.

**Comment by sbeitzel** [comment]
This might be interesting, but I see one hurdle to get over right away: calling out to an LLM with a request is almost certainly going to be an asynchronous action, so handling it inside a synchronous function gets problematic:

```swift
// let's use FoundationModels, skipping all the availability checks for clarity of example
public func merge(_ value: String, withSubordinate other: String, commonAncestor: String) throws -> String {
  var model = SystemLanguageModel.default
  let boundary = "---BOUNDARY+\(UUID().uuidString)---"
  let instructions = """
        Generate a three-way merge between a common ancestor string, a locally modified string,
        and a remotely modified string. The three strings are provided in order, separated by
        the boundary marker: \(boundary)
 """
  let session = LanguageModelSession(instructions: instructions)
  let prompt = " \(commonAncestor)\(boundary)\(value)\(boundary)\(other)"
  let mergedString = try await session.respond(to: prompt) // <--- 'async' call in a function that does not support concurrency
}
```

Even if we don't use Apple Intelligence but rather some other LLM (maybe an ollama instance running locally, or some custom model running on the corporate network, or whatever), that's still going to be an off-process call. I could see having a family of mergers that all do their processing asynchronously, so maybe a companion protocol `AsyncMerger` that accepts a callback, or produces a future or something...I'm not entirely sure what the right design choice is, here.

**Comment by sbeitzel** [comment]
Also, and possibly irrelevant but at least interesting: the example code above wouldn't get the right thing even aside from the asynchronous problem.

```swift
#Playground {
    if #available(macOS 26.0, *) {
        let boundary = "-+-"
        let value = "Fork Yeah!!!"
        let other = "Fork yeah"
        let commonAncestor = "Fork Yeah"
        let instructions = """
        The user's locale is \(Locale.current)
        Generate a three-way merge between a common ancestor string, a locally modified string,
        and a remotely modified string. The three strings are provided in order and
        delimited by a boundary marker, \(boundary)
        """
        let prompt = "\(boundary)\(commonAncestor)\(boundary)\(value)\(boundary)\(other)\(boundary)"
        let session = LanguageModelSession(instructions: instructions)
        let response = try await session.respond(to: prompt)
    } else {
        print("No model!!!")
    }
}
```
Running this playground, one sees that `response.content` contains a lot of stuff that isn't the merged string, and then at the end, the merge doesn't even do what we would want, instead it just concatenates.

> Sure! Here's the three-way merge of the strings:
>
> Fork Yeah
> Fork Yeah!!!
> Fork yeah
>
> The merged string is "Fork YeahFork Yeah!!!Fork yeah".

**Comment by drewmccormack** [comment]
You are right that there would need to be a new async pathway for merging. New async companion methods would need to be made available, and an AsyncMerger protocol or whatever. 

I think the LLM could be made to work, but it would take some playing with the instructions. I think the way they are presented above is too abstract. 

You would probably have something like:

```swift
let instructions = """
Initially there is this text: "Fork Yeah"
One person changes it to: "Fork Yeah!!!"
At the same time, another changes it to: "Fork yeah"

What would be a logical way to merge the two new pieces of text?

Include only the merged text in the response.
"""
```

I just threw this into Claude, and it worked:

```
> Initially there is this text: "Fork Yeah"
One person changes it to: "Fork Yeah!!!"
At the same time, another changes it to: "Fork yeah"

What would be a logical way to merge the two new pieces of text?

Include only the merged text in the response. 

⏺ Fork yeah!!!
```

Of course, Claude is powerful, but it shows that LLMs should be able to do it. Whether on-device is up to it right now, I don't know, but I would think it should be.
